lib/internal/fs/linux_watcher.js

```diff
@@ -0,0 +1,193 @@
+'use strict';
+
+const { EventEmitter } = require('events');
+const path = require('path');
+const { SafeMap, Symbol, StringPrototypeStartsWith } = primordials;
+const { validateObject } = require('internal/validators');
+const { kEmptyObject } = require('internal/util');
+const { ERR_FEATURE_UNAVAILABLE_ON_PLATFORM } = require('internal/errors');
+const { Dirent, UV_DIRENT_DIR } = require('internal/fs/utils');
+
+const kFSWatchStart = Symbol('kFSWatchStart');
+
+let internalSync;
+let internalPromises;
+
+function lazyLoadFsPromises() {
+  internalPromises ??= require('fs/promises');
+  return internalPromises;
+}
+
+function lazyLoadFsSync() {
+  internalSync ??= require('fs');
+  return internalSync;
+}
+
+function traverse(dir, files = new SafeMap()) {
+  const { readdirSync } = lazyLoadFsSync()
+
+  const filenames = readdirSync(dir, { withFileTypes: true });
```

Comment:
We might prefer using something asynchronous here, to avoid ending up blocking the thread if the directory contains too many items, wdyt?
```suggestion
  const filenames = await opendir(dir, { withFileTypes: true });
```

Reply:
I'm in favor of this change. But you should know that we have an inconsistency with OSX and Windows. `test-fs-watch-recursive.js` has the following if statement, because on Windows, the initial implementation is syncronous, but on macOS it's async. Therefore, I needed to add `await setTimeout` to all Linux tests. (The current implementation is async, even though traverse is sync.)

```
let interval;
if (common.isOSX || common.isLinux) {
  interval = setInterval(function() {
    fs.writeFileSync(filepathOne, 'world');
  }, 10);
} else {
  fs.writeFileSync(filepathOne, 'world');
}
```